lib/engine/game/g_1880/game.rb

```diff
@@ -203,9 +227,93 @@ def next_round!
             end
         end
 
+        def stock_round
+          G1880::Round::Stock.new(self, [
+            Engine::Step::DiscardTrain,
+            Engine::Step::Exchange,
+            Engine::Step::SpecialTrack,
+            Engine::Step::BuySellParShares,
+          ])
+        end
+
+        def operating_round(round_num)
+          Engine::Round::Operating.new(self, [
+            Engine::Step::Exchange,
+            Engine::Step::Track,
+            Engine::Step::Token,
+            G1880::Step::Route,
+            G1880::Step::Dividend,
+            Engine::Step::DiscardTrain,
+            Engine::Step::BuyTrain,
+          ], round_num: round_num)
+        end
+
+        def init_share_pool
+          G1880::SharePool.new(self)
+        end
+
+        def init_minors
+          game_minors.map { |minor| Minor.new(**minor) }
+        end
+
+        def setup
+          # reserve last 5 shares of each corp
+          @corporations.each do |c|
+            c.shares.last(5).each { |s| s.buyable = false }
+          end
+        end
+
         def p1
           company_by_id('P1')
         end
+
+        def event_float_30!
+          @log << "-- Event: #{EVENTS_TEXT['float_30'][1]} --"
```

Comment:
A minor comment is that it is possible to combine the code of these three events in one method, and then call the method from each event with suitable parameters, like 
def do_event_float(event_text_name, float_percent)
   ...
end

Reply:
There will be other events for those trains, which I think means i'll need to play around with the phase.rb code here:
https://github.com/tobymao/18xx/blob/526032884ca315ba29ac875e0edecc7212e3c407/lib/engine/phase.rb#L26

and get it send event_{event_name}! for all other events, and do_event_float for the float events.
That would be pretty messy. Or am i missing something?

Reply:
I just saw it as a refactoring opportunity as the code was more or less triplicated. But I did not check in details, and I do not know the final result, so it might be something to look for near the end. 
In this case it is also three simple methods so benefit would be low. The DRY principle is more to avoid code duplication to avoid errors caused by changes not applied in all instances, and also about readability. 

I can recommend "Clean code" books/article if you have not read them. (Not that I am saying that your code is not "clean" just as a general recommendation after doing a lot of code review professionally, and worked a lot with legacy code.)

Reply:
Agreed, I'm familiar with clean code and DRY. I just didn't see a way to do it here without introducing a lot more code, and if statements. I was asking if you saw a way to do so. Closing the comment.